File: test/benchmark/cases/std_sort/case.cpp
#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <vector>

auto repeat_count(int count) -> int
{
    if(count <= 2048) {
        return 512;
    }
    if(count <= 65536) {
        return 32;
    }
    return 2;
}

auto stable_requested() -> bool
{
    return std::getenv("CP_BENCH_STABLE") != nullptr;
}

auto value_at(int index, int count) -> int
{
    return static_cast<int>((static_cast<std::int64_t>(index) * 48271 + static_cast<std::int64_t>(count) * 17) % 1000003);
}

auto checked_checksum(std::vector<int> const& values) -> std::int64_t
{
    auto total = std::int64_t{ 0 };
    auto index = std::size_t{ 0 };
    while(index < values.size()) {
        if(index != 0 and values[index] < values[index - 1]) {
            return -1;
        }
        total = (total + static_cast<std::int64_t>(values[index]) * static_cast<std::int64_t>(index % 251 + 1)) % 1'000'000'007;
        index += 1;
    }
    return total;
}

auto main() -> int
{
    auto const* raw = std::getenv("CP_BENCH_INPUT");
    if(raw == nullptr) {
        return 1;
    }

    auto const count = std::atoi(raw);
    if(count <= 0) {
        return 1;
    }

    auto original = std::vector<int>{};
    original.reserve(static_cast<std::size_t>(count));
    auto index = 0;
    while(index < count) {
        original.push_back(value_at(index, count));
        index += 1;
    }

    auto total = std::int64_t{ 0 };
    auto round = 0;
    auto const repeats = repeat_count(count);
    auto const stable = stable_requested();
    while(round < repeats) {
        auto values = original;
        if(stable) {
            std::ranges::stable_sort(values);
        } else {
            std::ranges::sort(values);
        }
        auto const checksum = checked_checksum(values);
        if(checksum < 0) {
            return 2;
        }
        total = (total + checksum) % 1'000'000'007;
        round += 1;
    }

    return static_cast<int>(total % 251);
}

File: test/benchmark/cases/std_sort/case.rs
use std::env;

fn repeat_count(count: i32) -> i32 {
    if count <= 2048 {
        return 512;
    }
    if count <= 65536 {
        return 32;
    }
    2
}

fn stable_requested() -> bool {
    env::var("CP_BENCH_STABLE").is_ok()
}

fn value_at(index: i32, count: i32) -> i32 {
    (((index as i64) * 48_271 + (count as i64) * 17) % 1_000_003) as i32
}

fn checked_checksum(values: &[i32]) -> i64 {
    let mut total = 0i64;
    let mut index = 0usize;
    while index < values.len() {
        if index > 0 && values[index] < values[index - 1] {
            return -1;
        }
        total = (total + values[index] as i64 * (index % 251 + 1) as i64) % 1_000_000_007;
        index += 1;
    }
    total
}

fn main() {
    let count = env::var("CP_BENCH_INPUT")
        .ok()
        .and_then(|value| value.parse::<i32>().ok())
        .unwrap_or(1);
    if count <= 0 {
        std::process::exit(1);
    }

    let mut original = Vec::<i32>::with_capacity(count as usize);
    let mut index = 0i32;
    while index < count {
        original.push(value_at(index, count));
        index += 1;
    }

    let mut total = 0i64;
    let mut round = 0i32;
    let repeats = repeat_count(count);
    let stable = stable_requested();
    while round < repeats {
        let mut values = original.clone();
        if stable {
            values.sort();
        } else {
            values.sort_unstable();
        }
        let checksum = checked_checksum(&values);
        if checksum < 0 {
            std::process::exit(2);
        }
        total = (total + checksum) % 1_000_000_007;
        round += 1;
    }

    std::process::exit((total % 251) as i32);
}
